common/telemetry/env.go

```diff
@@ -0,0 +1,107 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package telemetry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	otelsdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.temporal.io/server/common/primitives"
+	"golang.org/x/exp/maps"
+)
+
+var (
+	unsupportedTraceExporter         = errors.New("unsupported OTEL exporter")
+	unsupportedTraceExporterProtocol = errors.New("unsupported OTEL exporter protocol")
+)
+
+const (
+	OtelServiceNameEnvKey                = "OTEL_SERVICE_NAME"
+	OtelTracesExporterTypesEnvKey        = "OTEL_TRACES_EXPORTER"
+	OtelTracesOtlpExporterType           = SpanExporterType("otlp")
+	OtelExporterOtlpTracesProtocolEnvKey = "OTEL_EXPORTER_OTLP_TRACES_PROTOCOL"
+	OtelExporterOtlpTracesGrcpProtocol   = "grpc"
+)
+
+type envVarLookup = func(string) (string, bool)
+
+// SupplementTraceExportersFromEnv adds any OTEL span exporters configured through environment variables
+// to a given set of exporters. It does not override an existing exporter with the same type.
+func SupplementTraceExportersFromEnv(
+	exporters map[SpanExporterType]otelsdktrace.SpanExporter,
+	envVars envVarLookup,
+) error {
+	exporterTypes, ok := envVars(OtelTracesExporterTypesEnvKey)
+	if !ok {
+		return nil
+	}
+
+	supplements := map[SpanExporterType]otelsdktrace.SpanExporter{}
+	for _, exporterType := range strings.Split(exporterTypes, ",") {
+		switch SpanExporterType(exporterType) {
+		case OtelTracesOtlpExporterType:
+			if _, exists := exporters[OtelTracesOtlpExporterType]; exists {
+				continue
+			}
+
+			// only grpc is supported; fail if user requests a different protocol
+			if protocol, exists := envVars(OtelExporterOtlpTracesProtocolEnvKey); exists {
+				isSupported := protocol == OtelExporterOtlpTracesGrcpProtocol
+				if !isSupported {
+					return fmt.Errorf("%w: %v=%v", unsupportedTraceExporterProtocol, OtelExporterOtlpTracesProtocolEnvKey, protocol)
+				}
+			}
+
+			// other OTEL configuration env variables are picked up automatically by the exporter itself
+			supplements[OtelTracesOtlpExporterType] = otlptracegrpc.NewUnstarted()
+		default:
+			return fmt.Errorf("%w: %v=%v", unsupportedTraceExporter, OtelTracesExporterTypesEnvKey, exporterType)
+		}
+	}
+	maps.Copy(exporters, supplements)
+	return nil
+}
```

Comment:
This is non-binding comment. 
Observation 1: If this method would return `supplements` instead of updating `exporterd` then at the call site your code would express what you are doing better. You would be able to write something like:
```
exporters := merge(SpanExporters(), ExportersFromEnv())
```
Observation 2:
To build supplements you iterate over exporterTypes, make sure that all of them of type `OtelTracesOtlpExporterType`, skip all those which are already in the exporters map for the rest, make sure that they are supported and add them to supplements.
Doing too many things in a single loop is a smell. Usually, it is a premature optimization where we try to iterate over a sequence only once. It prevents us from writing easy to read code. 

To state what the function is doing the code could look something like:
```
if !allExportersOfType(exporterTypes, OtelTracesOtlpType) {
    return fmt.Errorf(...)
}
toBeAddedExporters := dropIfPreset(exporters, exporterTypes)
if !allProtocolSupported(toBeAddedExporters) {
   return fmt.Errorf(...)
}

return utlis.FoldSlice(toBeAddedExporters, map[]{}, func(m map[], exporterType string) map[] { m[exporterType] = otlptravegrpc.NewUnstarted() })

```
I mixed styles here. On one hand I wrote dropIfPresent instead of utils.SliceFilter and used utils.FoldSlice instead of using some properly named helper function. I think for people not coming from FP helper funcitons are easier, but both produce very declarative code.

Reply:
re 1: I agree, that would be better. The main reason I wanted to keep the merge here was that it was easier to test here than at the call site.

re 2: the loop is doing a lot, that's true. The motivation wasn't optimization (since there will rarely be more than 1 item in the env variables). Let me see if I can simplify it.

Reply:
I;ve made some changes.